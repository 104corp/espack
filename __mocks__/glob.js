/* eslint-disable */
const glob = jest.genMockFromModule('glob');

let mockFiles = [];
function __setMockFiles(newMockFiles) {
  mockFiles = [];
  for (const index in newMockFiles) {
    if (/\.js$/i.test(newMockFiles[index])) {
      mockFiles.push(newMockFiles[index]);
    }
  }
}

function sync(src, options) {
  if (src === 'src/pages/**/_*.js') return [];
  return mockFiles;
}

glob.__setMockFiles = __setMockFiles;
glob.sync = sync;

module.exports = glob;
