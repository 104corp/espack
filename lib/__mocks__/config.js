const config = {
  options: {},
  setConfigOptions(options) {
    this.options = { ...this.options, ...options };
  },
  getConfigOptions() {
    return this.options;
  },
  ENV_PRODUCTION: 'production',
  ENV_DEVELOPMENT: 'development',
  SRC_FOLDER: 'src',
  PAGE_FOLDER: 'src/pages',
  EJS_FOLDER: 'src/pages',
  PUBLIC_FOLDER: 'public',
  BUILD_FOLDER: 'dist',
  TMP_BUILD_FOLDER: '.tmp/build',
  DEFAULT_TEMPLATE: '/template/_default.ejs',
  DEFAULT_JS: '/template/_about.js',
  INFO_CHUNK_NAME: '__about__',
  STATIC_ROOT: '.',
  CSS_ASSETS_PATH: '',
  VIEW_ROOT: '.',
  CONFIG_MERGE_JS: 'espack.config.js',
  BASE_TAG: '/',
  CLI_BUILD_TITLE: 'Bundling',
  CLI_DEV_TITLE: 'Starting',
  FONTS_FOLDER: 'fonts',
  IGNORE_PAGES_RULE: ['**/_*.js'],
  OUTPUT_FILE_NAME: {
    js: '[name].js?[chunkhash:8]',
    css: '[name].css?[contenthash:8]',
  },
};

config.setConfigOptions = config.setConfigOptions.bind(config);
config.getConfigOptions = config.getConfigOptions.bind(config);

module.exports = config;
