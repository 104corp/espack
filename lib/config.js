const config = {
  // shared variable
  options: {},

  setConfigOptions(options) {
    this.options = { ...this.options, ...options };
  },

  getConfigOptions() {
    return this.options;
  },


  // env production
  ENV_PRODUCTION: 'production',

  // env development
  ENV_DEVELOPMENT: 'development',

  // source folder
  SRC_FOLDER: 'src',

  // Default main program (entry) folder
  PAGE_FOLDER: 'src/pages',

  // ejs template source folder
  EJS_FOLDER: 'src/pages',

  // static folder, jpg, png, css, js...
  PUBLIC_FOLDER: 'public',

  // output root folder of production mode
  BUILD_FOLDER: 'dist',

  // development stage temporary release folder
  TMP_BUILD_FOLDER: '.tmp/build',

  // default template of ejs
  DEFAULT_TEMPLATE: `${__dirname}/template/_default.ejs`,

  // main program of default entry list, show information of espack
  DEFAULT_JS: `${__dirname}/template/_about.js`,

  // chunk name of default entry list page
  INFO_CHUNK_NAME: '__about__',

  // output root folder of static resource
  STATIC_ROOT: '.',

  // resource path for CSS
  CSS_ASSETS_PATH: '/',

  // output root folder of HTML
  VIEW_ROOT: '.',

  // config merge file
  CONFIG_MERGE_JS: 'espack.config.js',

  // <base href="<%= baseTag %>"> - base tag setting
  BASE_TAG: '/',

  // CLI Build Title
  CLI_BUILD_TITLE: 'Bundling',

  // CLI Dev Title
  CLI_DEV_TITLE: 'Starting',

  // output Fonts folder
  FONTS_FOLDER: 'fonts',

  // ignore pages rule
  IGNORE_PAGES_RULE: ['**/_*.js'],

  // output filename rule - production mode only
  // because chunkhash is not allowed under devServer hmr
  OUTPUT_FILE_NAME: {
    js: '[name].js?[chunkhash:8]',
    css: '[name].css?[contenthash:8]',
  },

};

config.setConfigOptions = config.setConfigOptions.bind(config);
config.getConfigOptions = config.getConfigOptions.bind(config);

module.exports = config;
