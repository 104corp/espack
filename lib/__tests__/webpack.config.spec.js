import webpackConfig from '../webpack.config';
import { setConfigOptions, OUTPUT_FILE_NAME } from '../config';

jest.mock('../config');
jest.mock('fs-extra');
jest.mock('glob');
jest.mock('webpack');
jest.mock('mini-css-extract-plugin');
jest.mock('uglifyjs-webpack-plugin');

describe('webpack.config.js', () => {
  beforeEach(() => {
    require('glob').__setMockFiles([ // eslint-disable-line no-underscore-dangle, global-require
      'src/pages/index.js',
    ]);
    setConfigOptions({
      workDirectory: '/',
      outputFileName: { ...OUTPUT_FILE_NAME },
    });
  });

  it('production config 驗證', () => {
    const config = webpackConfig({ dev: false, dropConsole: false });
    expect(config).toMatchSnapshot();
  });

  it('development config 驗證', () => {
    const config = webpackConfig({ dev: true });
    expect(config).toMatchSnapshot();
  });
});
