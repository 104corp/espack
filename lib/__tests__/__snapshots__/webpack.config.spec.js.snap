// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack.config.js development config 驗證 1`] = `
Object {
  "devtool": "source-map",
  "entry": Object {
    "./js/index": Array [
      "webpack-dev-server/client?http://localhost:8888/",
      "webpack/hot/dev-server",
      "./src/pages/index.js",
    ],
    "__about__": Array [
      "webpack-dev-server/client?http://localhost:8888/",
      "webpack/hot/dev-server",
      "/template/_about.js",
    ],
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "loader": "ejs-loader",
        "test": /\\\\\\.ejs\\$/,
      },
      Object {
        "exclude": /\\(node_modules\\|bower_components\\)/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Object {
          "loader": "babel-loader",
          "options": Object {
            "cacheDirectory": true,
            "plugins": Array [
              "lodash",
            ],
          },
        },
      },
      Object {
        "test": /\\\\\\.\\(jpe\\?g\\|png\\|gif\\|svg\\)\\$/i,
        "use": Array [
          Object {
            "loader": "url-loader",
            "options": Object {
              "limit": 10000,
              "name": [Function],
              "publicPath": [Function],
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.\\(woff2\\?\\|ttf\\|eot\\|otf\\)\\$/,
        "use": Array [
          Object {
            "loader": "file-loader",
            "options": Object {
              "name": [Function],
              "publicPath": [Function],
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.\\(css\\|s\\[ac\\]ss\\)\\$/,
        "use": Array [
          "css-hot-loader",
          [Function],
          Object {
            "loader": "css-loader",
            "options": Object {
              "alias": Object {
                "@": "/",
              },
              "minimize": true,
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "ident": "postcss",
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": true,
            },
          },
          Object {
            "loader": "resolve-url-loader",
            "options": Object {
              "root": "/node_modules",
              "sourceMap": true,
            },
          },
          Object {
            "loader": "sass-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.less\\$/,
        "use": Array [
          "css-hot-loader",
          [Function],
          Object {
            "loader": "css-loader",
            "options": Object {
              "alias": Object {
                "@": "/",
              },
              "minimize": true,
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "ident": "postcss",
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": true,
            },
          },
          Object {
            "loader": "less-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "splitChunks": Object {
      "cacheGroups": Object {
        "default": false,
      },
    },
  },
  "output": Object {
    "chunkFilename": "[name].js",
    "filename": "[name].js?[chunkhash:8]",
    "path": "/dist",
    "publicPath": "",
  },
  "plugins": Array [
    Object {},
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunkList": Array [
          Object {
            "link": "./index.html",
            "text": "./index.html",
          },
        ],
        "chunks": Array [
          "__about__",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": true,
        "meta": Object {},
        "minify": false,
        "showErrors": true,
        "template": "/template/_default.ejs",
        "templateParameters": [Function],
        "title": "Webpack App",
        "xhtml": false,
      },
    },
    HtmlWebpackPlugin {
      "options": Object {
        "baseTag": "/",
        "cache": true,
        "chunks": Array [
          "./js/index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "./index.html",
        "hash": false,
        "inject": "body",
        "meta": Object {},
        "minify": Object {
          "collapseWhitespace": false,
          "minifyCSS": false,
          "minifyJS": false,
          "preserveLineBreaks": false,
          "removeComments": false,
        },
        "showErrors": true,
        "template": "./src/pages/index.ejs",
        "templateParameters": [Function],
        "title": "Webpack App",
        "xhtml": false,
      },
    },
    Object {
      "filename": "[name].css?[contenthash:8]",
    },
    LodashWebpackPlugin {},
    CSSExtractRenamePlugin {},
  ],
  "resolve": Object {
    "extensions": Array [
      ".web.js",
      ".js",
      ".json",
      ".web.jsx",
      ".jsx",
      ".css",
      ".scss",
      ".sass",
      ".less",
    ],
    "modules": Array [
      "bower_components",
      "node_modules",
    ],
  },
}
`;

exports[`webpack.config.js production config 驗證 1`] = `
Object {
  "devtool": false,
  "entry": Object {
    "./js/index": Array [
      "./src/pages/index.js",
    ],
  },
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "loader": "ejs-loader",
        "test": /\\\\\\.ejs\\$/,
      },
      Object {
        "exclude": /\\(node_modules\\|bower_components\\)/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Object {
          "loader": "babel-loader",
          "options": Object {
            "cacheDirectory": true,
            "plugins": Array [
              "lodash",
            ],
          },
        },
      },
      Object {
        "test": /\\\\\\.\\(jpe\\?g\\|png\\|gif\\|svg\\)\\$/i,
        "use": Array [
          Object {
            "loader": "url-loader",
            "options": Object {
              "limit": 10000,
              "name": [Function],
              "publicPath": [Function],
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.\\(woff2\\?\\|ttf\\|eot\\|otf\\)\\$/,
        "use": Array [
          Object {
            "loader": "file-loader",
            "options": Object {
              "name": [Function],
              "publicPath": [Function],
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.\\(css\\|s\\[ac\\]ss\\)\\$/,
        "use": Array [
          "css-hot-loader",
          [Function],
          Object {
            "loader": "css-loader",
            "options": Object {
              "alias": Object {
                "@": "/",
              },
              "minimize": true,
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "ident": "postcss",
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": true,
            },
          },
          Object {
            "loader": "resolve-url-loader",
            "options": Object {
              "root": "/node_modules",
              "sourceMap": true,
            },
          },
          Object {
            "loader": "sass-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.less\\$/,
        "use": Array [
          "css-hot-loader",
          [Function],
          Object {
            "loader": "css-loader",
            "options": Object {
              "alias": Object {
                "@": "/",
              },
              "minimize": true,
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "ident": "postcss",
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": true,
            },
          },
          Object {
            "loader": "less-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "minimizer": Array [
      Object {
        "parallel": true,
        "sourceMap": false,
        "uglifyOptions": Object {
          "compress": Object {
            "dead_code": true,
            "drop_console": false,
            "warnings": false,
          },
          "ie8": true,
          "output": Object {
            "comments": false,
          },
        },
      },
    ],
    "splitChunks": Object {
      "cacheGroups": Object {
        "default": false,
      },
    },
  },
  "output": Object {
    "chunkFilename": "[name].js",
    "filename": "[name].js?[chunkhash:8]",
    "path": "/dist",
    "publicPath": "",
  },
  "plugins": Array [
    Object {},
    HtmlWebpackPlugin {
      "options": Object {
        "baseTag": "/",
        "cache": true,
        "chunks": Array [
          "./js/index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "./index.html",
        "hash": false,
        "inject": "body",
        "meta": Object {},
        "minify": Object {
          "collapseWhitespace": false,
          "minifyCSS": false,
          "minifyJS": false,
          "preserveLineBreaks": false,
          "removeComments": false,
        },
        "showErrors": true,
        "template": "./src/pages/index.ejs",
        "templateParameters": [Function],
        "title": "Webpack App",
        "xhtml": false,
      },
    },
    Object {
      "filename": "[name].css?[contenthash:8]",
    },
    LodashWebpackPlugin {},
    CSSExtractRenamePlugin {},
  ],
  "resolve": Object {
    "extensions": Array [
      ".web.js",
      ".js",
      ".json",
      ".web.jsx",
      ".jsx",
      ".css",
      ".scss",
      ".sass",
      ".less",
    ],
    "modules": Array [
      "bower_components",
      "node_modules",
    ],
  },
}
`;
